web/src/js/components/eventlog.js

```diff
@@ -1,115 +1,151 @@
+
+    componentDidUpdate() {
+        this.onViewportUpdate();
+    }
+
+    componentWillReceiveProps(nextProps) {
         if (nextProps.filter !== this.props.filter) {
-            this.props.filter = nextProps.filter; // Dirty: Make sure that view filter sees the update.
-            this.state.view.recalculate();
+            this.view.recalculate(
+                entry => nextProps.filter[entry.level]
+            );
         }
-    },
-    getDefaultProps: function () {
-        return {
-            rowHeight: 45,
-            rowHeightMin: 15,
-            placeholderTagName: "div"
-        };
-    },
-    renderRow: function (elem) {
-        return <LogMessage key={elem.id} entry={elem}/>;
-    },
-    render: function () {
-        var entries = this.state.view.list;
-        var rows = this.renderRows(entries);
-
-        return <pre onScroll={this.onScroll}>
-            { this.getPlaceholderTop(entries.length) }
-            {rows}
-            { this.getPlaceholderBottom(entries.length) }
-        </pre>;
     }
-});
 
-var ToggleFilter = React.createClass({
-    toggle: function (e) {
-        e.preventDefault();
-        return this.props.toggleLevel(this.props.name);
-    },
-    render: function () {
-        var className = "label ";
-        if (this.props.active) {
-            className += "label-primary";
-        } else {
-            className += "label-default";
+    onViewportUpdate() {
+        const viewport = ReactDOM.findDOMNode(this);
+
+        const vScroll = calcVScroll({
+            itemCount: this.state.entries.length,
+            rowHeight: this.props.rowHeight,
+            viewportTop: viewport.scrollTop,
+            viewportHeight: viewport.offsetHeight,
+            itemHeights: this.state.entries.map(entry => this.heights[entry.id]),
+        });
+
+        if (!shallowEqual(this.state.vScroll, vScroll)) {
+            this.setState({ vScroll });
+        }
+    }
+
+    onChange() {
+        this.setState({ entries: this.view.list });
+    }
+
+    setHeight(id, ref) {
+        if (ref) {
+            const height = ReactDOM.findDOMNode(ref).offsetHeight;
+            if (this.heights[id] !== height) {
+                this.heights[id] = height;
+                this.onViewportUpdate();
+            }
         }
+    }
+
+    getIcon(level) {
+        return { web: "html5", debug: "bug" }[level] || "info";
+    }
+
+    render() {
+        const vScroll = this.state.vScroll;
+        const entries = this.state.entries.slice(vScroll.start, vScroll.end);
+
         return (
-            <a
-                href="#"
-                className={className}
-                onClick={this.toggle}>
-                {this.props.name}
-            </a>
+            <pre onScroll={this.onViewportUpdate}>
+                <div style={{ height: vScroll.paddingTop }}></div>
+                {entries.map((entry, index) => (
+                    <div key={entry.id} ref={this.setHeight.bind(this, entry.id)}>
```

Comment:
@mhils Do you mean a stateless component? We cannot have refs on stateless components. So if we do that we should either abandon the variable height mechanism or add some stuffs to pass refs back to the parent. :confused: